x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTrunc.java

```diff
@@ -0,0 +1,158 @@
+                    return datePart;
+                }
+            }
+            return null;
+        }
+    }
+
+
+    private final ZoneId zoneId;
+
+    public DateTrunc(Source source, Expression truncateTo, Expression timestamp, ZoneId zoneId) {
+        super(source, truncateTo, timestamp);
+        this.zoneId = zoneId;
+    }
+
+    @Override
+    public DataType dataType() {
+        return right().dataType();
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isString(left(), sourceText(), Expressions.ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (left().foldable() && DatePart.resolveTruncate((String) left().fold()) == null) {
+            return new TypeResolution(format(null, "first argument of [{}] must be one of {} or their aliases, found value [{}]",
+                sourceText(),
+                DatePart.values(),
+                Expressions.name(left())));
+        }
+        resolution = isDate(right(), sourceText(), Expressions.ParamOrdinal.SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    protected BinaryScalarFunction replaceChildren(Expression newTruncateTo, Expression newTimestamp) {
+        return new DateTrunc(source(), newTruncateTo, newTimestamp, zoneId);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, DateTrunc::new, left(), right(), zoneId);
+    }
+
+    @Override
+    protected Pipe makePipe() {
+        return new DateTruncPipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), zoneId);
+    }
+
+    @Override
+    public Nullability nullable() {
+        return Nullability.TRUE;
+    }
+
+    @Override
+    public Object fold() {
+        return process(left().fold(), right().fold(), zoneId);
+    }
+
+    @Override
+    protected ScriptTemplate asScriptFrom(ScriptTemplate leftScript, ScriptTemplate rightScript) {
+        return new ScriptTemplate(
+            formatTemplate("{sql}.dateTrunc(" + leftScript.template() + "," + rightScript.template()+ ",{})"),
+            paramsBuilder()
+                .script(leftScript.params())
+                .script(rightScript.params())
+                .variable(zoneId.getId())
+                .build(),
+            dataType());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+        DateTrunc dateTrunc = (DateTrunc) o;
+        return Objects.equals(zoneId, dateTrunc.zoneId);
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
Small nitpick, generally within the code hashcode appears before equals (though not consistently).

Reply:
I think it's quite a mixture, but can change.